File: packages/emitter/testcases/common/types/functor-patterns/MaybeMonad.ts
export class Functor<T> {
  map<U>(fn: (x: T) => U): Functor<U> {
    throw new Error("Not implemented");
  }
}

export class Maybe<T> extends Functor<T> {
  private value: T | null;

  constructor(value: T | null) {
    super();
    this.value = value;
  }

  static just<T>(value: T): Maybe<T> {
    return new Maybe(value);
  }

  static nothing<T>(): Maybe<T> {
    return new Maybe<T>(null);
  }

  map<U>(fn: (x: T) => U): Maybe<U> {
    if (this.value === null) {
      return Maybe.nothing<U>();
    }
    return Maybe.just(fn(this.value));
  }

  flatMap<U>(fn: (x: T) => Maybe<U>): Maybe<U> {
    if (this.value === null) {
      return Maybe.nothing<U>();
    }
    return fn(this.value);
  }

  getOrElse(defaultValue: T): T {
    return this.value !== null ? this.value : defaultValue;
  }
}

File: packages/emitter/testcases/common/expected/types/functor-patterns/MaybeMonad.cs
namespace TestCases.common.types.functorpatterns
{
    public class Functor<T>
        where T : class
    {
        public virtual Functor<U> map<U>(global::System.Func<T, U> fn)
            where U : class
        {
            throw "Not implemented";
        }
    }

    public static class Maybe
    {
        public static Maybe<T> just<T>(T value)
            where T : class
        {
            return new Maybe<T>(value);
        }

        public static Maybe<T> nothing<T>()
            where T : class
        {
            return new Maybe<T>(default);
        }
    }

    public class Maybe<T> : Functor<T>
        where T : class
    {
        private T? __private_value;

        public Maybe(T? value) : base()
        {
            this.__private_value = value;
        }

        public override Maybe<U> map<U>(global::System.Func<T, U> fn)
            where U : class
        {
            if (this.__private_value == null)
            {
                return Maybe.nothing<U>();
            }
            return Maybe.just(fn(this.__private_value));
        }

        public Maybe<U> flatMap<U>(global::System.Func<T, Maybe<U>> fn)
            where U : class
        {
            if (this.__private_value == null)
            {
                return Maybe.nothing<U>();
            }
            return fn(this.__private_value);
        }

        public T getOrElse(T defaultValue)
        {
            return this.__private_value != null ? this.__private_value : defaultValue;
        }
    }
}
